plugins/default/src/Ide/Plugin/ModuleName.hs

```diff
@@ -0,0 +1,162 @@
+
+-- |Plugin descriptor
+descriptor :: PluginId -> PluginDescriptor
+descriptor plId =
+  (defaultPluginDescriptor plId)
+    { pluginId = plId,
+        pluginCodeLensProvider = Just codeLens
+       ,pluginCommands = [PluginCommand editCommandName editCommandName editCmd]
+        -- pluginCodeActionProvider = Just codeAction
+    }
+
+-- | Generate code lenses
+codeLens :: LspFuncs c -> IdeState -> PluginId -> CodeLensParams -> IO (Either a2 (List CodeLens))
+codeLens lsp state pluginId (CodeLensParams (TextDocumentIdentifier uri) _) = do
+  pid <- getPid
+  actions (asCodeLens (mkLspCmdId pid pluginId editCommandName)) lsp state uri
+
+-- | Generate code actions.
+-- NOTE: Not invoked on an empty module (but codeLens is, why?)
+codeAction :: LspFuncs c -> IdeState -> p1 -> TextDocumentIdentifier -> p2 -> p3 -> IO (Either a (List CAResult))
+codeAction lsp state _plId (TextDocumentIdentifier uri) _range _ = actions asCodeAction lsp state uri
+
+-- Copied from "Ide.Plugin"
+mkLspCmdId :: T.Text -> PluginId -> CommandId -> T.Text
+mkLspCmdId pid (PluginId plid) (CommandId cid)
+  = pid <> ":" <> plid <> ":" <> cid
+
+editCommandName :: IsString p => p
+editCommandName = "edit"
+
+-- | Generic command to apply a group of edits
+editCmd :: CommandFunction WorkspaceEdit
+editCmd _lf _ide workspaceEdits = return (Right Null, Just $ (WorkspaceApplyEdit,ApplyWorkspaceEditParams workspaceEdits))
+
+-- | Required actions (actually, at most one) that can be converted to either CodeLenses or CodeActions
+actions :: Show a1 => (Action -> a1) -> LspFuncs c -> IdeState -> Uri -> IO (Either a2 (List a1))
+actions convert lsp state uri = do
+    let Just nfp = uriToNormalizedFilePath $ toNormalizedUri uri
+    let Just fp = uriToFilePath' uri
+
+    contents <- liftIO $ getVirtualFileFunc lsp $ toNormalizedUri uri
+    let emptyModule = maybe True ((==0) . T.length . T.strip . virtualFileText) contents
+
+    correctNameMaybe <- pathModuleName state nfp fp
+    statedNameMaybe <- codeModuleName state nfp
+
+    let act = Action uri
+    let actions = case (correctNameMaybe,statedNameMaybe) of
+            (Just correctName,Just (nameRange,statedName)) | correctName /= statedName -> [convert $ act nameRange ("Set module name to " <> correctName)  correctName]
+            (Just correctName,_) | emptyModule -> let code = T.unwords ["module",correctName,"where\n"] in [convert $ act (Range (Position 0 0) (Position 0 0)) code code]
+            _ -> []
+
+    out ["actions",show actions]
+    pure . Right . List $ actions
+
+-- | The module name, as derived by the position of the module in its source directory
+pathModuleName :: IdeState -> NormalizedFilePath -> String -> IO (Maybe Text)
+pathModuleName state nfp fp =  do
+    session :: HscEnvEq <- runAction "ModuleName.ghcSession" state $ use_ GhcSession nfp
+    paths <- evalGhcEnv (hscEnv session) $ do
+        env <- getSession
+        let df = ic_dflags . hsc_IC  $ env
+        return $ importPaths df
```

Comment:
Instead of looking at the interactive context, extract the import paths directly from the HscEnvEq

Reply:
I pushed this fix:

paths <- evalGhcEnv (hscEnvWithImportPaths session) $ importPaths <$> getSessionDynFlags

is this what you meant?

Reply:
or even simpler
```
let paths = envImportPaths session
```

Reply:
Great, even simpler, so:

let paths = fromMaybe [] $ envImportPaths session

But are they strictly equivalent?

I don't quite grok the meaning of:
-- ^ If Just, import dirs originally configured in this env
--   If Nothing, the env import dirs are unaltered

Reply:
Right, that's for the no cradle case. Your previous version is preferable then: 
```
paths <- evalGhcEnv (hscEnvWithImportPaths session) $ importPaths <$> getSessionDynFlags
```